### PR TITLE
 [Improvement][API] ignore noNodeException when get worker groups

### PR DESCRIPTION
**Pull request by felix-thinkingdata**
#4109  ignore noNodeException when get worker groups

### CODE REVIEW

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/WorkerGroupService.java [hunk 1]
```diff
@@ -120,30 +121,46 @@
...
-        String workerPath = zookeeperCachedOperator.getZookeeperConfig().getDsRoot()+"/nodes" +"/worker";
-        List<String> workerGroupList = zookeeperCachedOperator.getChildrenKeys(workerPath);
+
+        String workerPath = zookeeperCachedOperator.getZookeeperConfig().getDsRoot() + "/nodes" + "/worker";
+        List<WorkerGroup> workerGroups = new ArrayList<>();
+        List<String> workerGroupList;
+        try {
+            workerGroupList = zookeeperCachedOperator.getChildrenKeys(workerPath);
+        } catch (Exception e) {
+            if (e.getMessage().contains(noNodeExceptionRegex)) {
```
Comment (CalvinKirs) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/WorkerGroupService.java:138]: It's best to pull it out here, and the code will be more readable.
抽离这里的判断

**Comment by CalvinKirs** [comment]
hi, do you guys have any ideas @lenboo @chenxingchun

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/WorkerGroupService.java (hunk 1 above)
Comment (felix-thinkingdata) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/WorkerGroupService.java:138]: I need to deal with two results, which is not easy to deal with。

这块要处理两个结果，抽取方法，不太好处理。

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/WorkerGroupService.java [hunk 2]
```diff
@@ -120,30 +121,46 @@
...
+     * get worker groups
      *
      * @param isPaging whether paging
      * @return WorkerGroup list
      */
     private List<WorkerGroup> getWorkerGroups(boolean isPaging) {
-        String workerPath = zookeeperCachedOperator.getZookeeperConfig().getDsRoot()+"/nodes" +"/worker";
-        List<String> workerGroupList = zookeeperCachedOperator.getChildrenKeys(workerPath);
+
+        String workerPath = zookeeperCachedOperator.getZookeeperConfig().getDsRoot() + "/nodes" + "/worker";
```
Comment (lenboo) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/WorkerGroupService.java:133]: i think it should be a constant variable  for worker path.

Comment (felix-thinkingdata) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/WorkerGroupService.java]: complete

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/WorkerGroupService.java [hunk 3]
```diff
@@ -120,30 +122,46 @@
...
+        try {
+            workerGroupList = zookeeperCachedOperator.getChildrenKeys(workerPath);
+        } catch (Exception e) {
+            if (e.getMessage().contains(NO_NODE_EXCEPTION_REGEX)) {
+                if (isPaging) {
+                    return workerGroups;
+                } else {
+                    //ignore noNodeException return Default
+                    WorkerGroup wg = new WorkerGroup();
+                    wg.setName("default");
```
Comment (lenboo) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/WorkerGroupService.java:145]: so do the "default" = DEFAULT_WORKER_GROUP

Comment (felix-thinkingdata) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/WorkerGroupService.java:145]: ok

**Review by lenboo — APPROVED** [review]
+1

**Review by CalvinKirs — APPROVED** [review]
+1

**Pull request merged**